File: 0031-next-permutation/0031-next-permutation.py
class Solution:
    def nextPermutation(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        n = len(nums)
        index = -1
        
        for i in range(n - 2, -1, -1):
            if nums[i] < nums[i + 1]:
                index = i
                break
        
        if index == -1:
            nums[:] = nums[::-1]
        else:

            min = float("inf")
            minIndex = 0
            for i in range(index + 1, n):
                if nums[i] < min and nums[i] > nums[index]:
                    min = nums[i]
                    minIndex = i
            print(index, minIndex)
            nums[index], nums[minIndex] = nums[minIndex], nums[index]
            
            nums[:] = nums[:index + 1] + sorted(nums[index + 1: ])

File: 0031-next-permutation/0031-next-permutation.java
class Solution {
    public void nextPermutation(int[] nums) {
        int pivot = -1;
        int n = nums.length;

        for (int i = n - 2; i >= 0; i--) {
            if (nums[i] < nums[i + 1]) {
                pivot = i;
                break;
            }
            
        }

        if (pivot >= 0) {
            int j = n - 1;

            while ( nums[j] <= nums[pivot]) {
                j--;
            }

            swap(nums, pivot, j);
        }

        System.out.print(pivot);

        reverse(nums, pivot + 1, n - 1);
    }

    private void swap(int[] nums, int i, int j) {
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }

    private void reverse(int[] nums, int l, int r) {
        while (l < r) {
            swap(nums, l++, r--);
        }
    }
}